neuraxle/hyperparams/distributions.py

```diff
@@ -468,6 +537,41 @@ def rvs(self) -> float:
             result = max(result, self.hard_clip_min)
         return float(result)
 
+    def pdf(self, x) -> float:
+        """
+        Calculate the Normal probability distribution value at position `x`.
+        :param x: value where the probability distribution function is evaluated.
+        :return: value of the probability distribution function.
+        """
+
+        if self.hard_clip_min is not None and (x < self.hard_clip_min):
+            return 0.
+
+        if self.hard_clip_min is not None and (x == self.hard_clip_min):
+            return norm.cdf(x, loc=self.mean, scale=self.std)
+
+        if self.hard_clip_max is not None and (x == self.hard_clip_max):
+            return 1 - norm.cdf(x, loc=self.mean, scale=self.std)
+
+        if self.hard_clip_max is not None and (x > self.hard_clip_max):
+            return 0.
+
+        return 1 / (self.std * math.sqrt(2 * math.pi)) * math.exp(-(x - self.mean) ** 2 / (2 * self.std ** 2))
+
+    def cdf(self, x) -> float:
+        """
+        Calculate the Normal cumulative distribution value at position `x`.
+        :param x: value where the cumulative distribution function is evaluated.
+        :return: value of the cumulative distribution function.
+        """
+        if self.hard_clip_min is not None and (x < self.hard_clip_min):
+            return 0.
+
+        if self.hard_clip_max is not None and (x >= self.hard_clip_max):
+            return 1.
+
+        return norm.cdf(x, loc=self.mean, scale=self.std)
```

Comment:
Note: I might be wrong about the rescaling, but I somehow feel that things should work this way. I didn't do a thorough math proof in my head, I suggest this as a feeling that a CDF should work this way, but my assumption from this feeling might be false.

Reply:
If I am not mistaken the clip function gives everything beyond a limit the value of this limit. For exemple if you clip at x = -2. Any value x = -3, -4, 5, will take the value -2. The proba of getting -3, -4, -5, -6 before clipping is still a gaussian, but at the end it all becomes -2. So you compress all the probabilities to -2 that are all being some. This summing, it is equivalent to the cdf at x=-2.  So this seams strange, but the pdf at x=-2, is equivalent to the cdf at x=-2. The cdf at x=-2 is unchanged however as the cdf, as even if it is clip, everything < x=-2 is 0. It is then only the pdf at x=-2 that counts, which is in fact the cdf. This works for the lower limits.

For the upper limits, it is the same principle, but the cdf will be 1.0 at x = upper_limit, because anything beyond has zero proba to have that value.

If you do a rescale, as you suggestion you will still have a gaussian so non-zero probablement beyond your limit, which is not what clip does, if I am right?

I will plot the cdf and pdf for comparaison so you can see it.

Reply:
@Eric2Hamel I understand, that's fine, thanks, you don't need to plot. 

I confused clipping with retrying to sample the out-of-bounds values.

Reply:
I wonder if we should instead re-sample the values out of bounds until we get something in the bounds (with a while True but by making it safe with a max number of retries and then picking the mininum).

Reply:
It depends on the comportment that you want to have. With clipping, the problem is that you are compressing all probababilité to one value which is not always great and give you more chance to obtained these values. Instead, re-sampling, will more keep you the same probability curve but willl increase the probability curve everywhere within your limit). This will ensure that values will be sample similar to a gaussian probability within your limit.

Reply:
I think we could accept the PR like it is for now and later possibly change how it works to instead use re-sampling, which seems better considering it doesn't increase the chance to get exactly what is at the limit, but rather increase the chance of getting everything else in general within the limit.

Reply:
Resolving to issue #110